src/MongoDB.Driver.Core/Core/Misc/SemanticVersion.cs

```diff
@@ -313,5 +321,159 @@ public static bool TryParse(string value, out SemanticVersion result)
+            private readonly int _patch;
+            private readonly int? _releaseCandidate;
+
+            // constructors
+            internal ServerVersion(int major, int minor, int patch, int? releaseCandidate, int? internalBuild, string commitHash)
+            {
+                _major = Ensure.IsGreaterThanOrEqualToZero(major, nameof(major));
+                _minor = Ensure.IsGreaterThanOrEqualToZero(minor, nameof(minor));
+                _patch = Ensure.IsGreaterThanOrEqualToZero(patch, nameof(patch));
+                _releaseCandidate = releaseCandidate; // can be null
+                _internalBuild = internalBuild; // can be null
+                _commitHash = commitHash; // can be null
+            }
+
+            // public methods
+            public int CompareTo(ServerVersion other)
+            {
+                if (other == null)
+                {
+                    return 1;
+                }
+
+                var result = _major.CompareTo(other._major);
+                if (result != 0)
+                {
+                    return result;
+                }
+
+                result = _minor.CompareTo(other._minor);
+                if (result != 0)
+                {
+                    return result;
+                }
+
+                result = _patch.CompareTo(other._patch);
+                if (result != 0)
+                {
+                    return result;
+                }
+
+                if (_releaseCandidate != null || other._releaseCandidate != null)
+                {
+                    if (_releaseCandidate == null)
+                    {
+                        return 1;
+                    }
+                    if (other._releaseCandidate == null)
+                    {
+                        return -1;
+                    }
+
+                    result = _releaseCandidate.Value.CompareTo(other._releaseCandidate.Value);
+                    if (result != 0)
+                    {
+                        return result;
+                    }
+                }
+
+                if (_internalBuild != null || other._internalBuild != null)
+                {
+                    if (_internalBuild == null)
+                    {
+                        return -1;
+                    }
+                    if (other._internalBuild == null)
+                    {
+                        return 1;
+                    }
+
+                    result = _internalBuild.Value.CompareTo(other._internalBuild.Value);
+                    if (result != 0)
+                    {
+                        return result;
+                    }
+                }
+
+                if (_commitHash == null && other._commitHash == null)
+                {
+                    return 0;
+                }
+                if (_commitHash == null)
+                {
+                    return -1;
+                }
+
+                return _commitHash.CompareTo(other._commitHash);
```

Comment:
Lines 423-342 can be replaced with:

    return 0; // or throw

It makes no sense to compare commit hashes.

Reply:
I've added along with a bit more extended `null` check.
Done.